packages/opentelemetry-plugin-user-interaction/src/userInteraction.ts

```diff
@@ -0,0 +1,457 @@
+    if (span) {
+      const uiMem = this._uiMem.get(span);
+      if (uiMem) {
+        span.end(endTime);
+        this._uiMem.delete(span);
+      }
+    }
+  }
+
+  /**
+   * implements patch function
+   */
+  protected patch() {
+    const _window = window as any;
+    this._logger.debug(
+      'applying patch to',
+      this.moduleName,
+      this.version,
+      'zone:',
+      !!_window.Zone
+    );
+    if (this.isZoneAvailable()) {
+      const ZoneWithPrototype = (_window.Zone as unknown) as ZoneTypeWithPrototype;
+      if (isWrapped(ZoneWithPrototype.prototype.runTask)) {
+        shimmer.unwrap(ZoneWithPrototype.prototype, 'runTask');
+        this._logger.debug('removing previous patch from method runTask');
+      }
+      if (isWrapped(ZoneWithPrototype.prototype.scheduleTask)) {
+        shimmer.unwrap(ZoneWithPrototype.prototype, 'scheduleTask');
+        this._logger.debug('removing previous patch from method scheduleTask');
+      }
+      if (isWrapped(ZoneWithPrototype.prototype.cancelTask)) {
+        shimmer.unwrap(ZoneWithPrototype.prototype, 'cancelTask');
+        this._logger.debug('removing previous patch from method cancelTask');
+      }
+
+      this._zonePatched = true;
+      shimmer.wrap(
+        ZoneWithPrototype.prototype,
+        'runTask',
+        this._patchZoneRunTask()
+      );
+      shimmer.wrap(
+        ZoneWithPrototype.prototype,
+        'scheduleTask',
+        this._patchZoneScheduleTask()
+      );
+      shimmer.wrap(
+        ZoneWithPrototype.prototype,
+        'cancelTask',
+        this._patchZoneCancelTask()
+      );
+    } else {
+      this._zonePatched = false;
+      if (isWrapped(HTMLElement.prototype.addEventListener)) {
+        shimmer.unwrap(HTMLElement.prototype, 'addEventListener');
+        this._logger.debug(
+          'removing previous patch from method addEventListener'
+        );
+      }
+      shimmer.wrap(
+        HTMLElement.prototype,
+        'addEventListener',
+        this._patchElement()
+      );
+    }
+
+    this._patchHistoryApi();
+    return this._moduleExports;
+  }
+
+  /**
+   * implements unpatch function
+   */
+  protected unpatch() {
+    const _window = window as any;
+    this._logger.debug(
+      'removing patch from',
+      this.moduleName,
+      this.version,
+      'zone:',
+      !!_window.Zone
+    );
+    if (this.isZoneAvailable() && this._zonePatched) {
+      const ZoneWithPrototype = (_window.Zone as unknown) as ZoneTypeWithPrototype;
+      shimmer.unwrap(ZoneWithPrototype.prototype, 'runTask');
+      shimmer.unwrap(ZoneWithPrototype.prototype, 'scheduleTask');
+      shimmer.unwrap(ZoneWithPrototype.prototype, 'cancelTask');
+    } else {
+      shimmer.unwrap(HTMLElement.prototype, 'addEventListener');
+    }
+    this._unpatchHistoryApi();
+  }
+
+  /**
+   * returns true if zone.js is available
+   */
+  isZoneAvailable() {
+    const _window = window as any;
+    return _window.Zone;
```

Comment:
What is the purpose of the `_window` const here?

Reply:
Zone is not part of window object, you can't reference without having some typing - typescript will complain.

Reply:
Can this be simplified to something like `return (window as WindowWithZone).Zone;`? Assuming you create a `WindowWithZone` interface so we don't need to use `any` here.

Reply:
could we also

```typescript
declare global {
    interface Window { Zone: Zone; }
}
```